quixstreams/state/windows.py

```diff
@@ -0,0 +1,355 @@
+import functools
+from datetime import timedelta
+from typing_extensions import TypedDict
+from typing import Any, Optional, Callable, TypeVar, Tuple
+
+from quixstreams.state import StateStoreManager
+from quixstreams.state.rocksdb.windowed.store import WindowedTransactionState
+from quixstreams.context import (
+    message_context,
+    message_key,
+)
+
+
+class WindowResult(TypedDict):
+    start: float
+    end: float
+    value: Any
+
+
+T = TypeVar("T")
+WindowedAggregateFunc = Callable[
+    [float, float, float, T, WindowedTransactionState], Any
+]
+WindowedDataFrameFunc = Callable[
+    [T, WindowedTransactionState, Optional[Callable]],
+    Tuple[list[WindowResult], list[WindowResult]],
+]
+StreamingDataFrame = TypeVar("StreamingDataFrame")
+
+
+class TumblingWindowDefinition:
+    def __init__(
+        self,
+        duration: float | timedelta,
+        grace: float | timedelta,
+        dataframe: StreamingDataFrame,
+        name: Optional[str] = None,
+    ):
+        self._duration = (
+            duration.total_seconds() if isinstance(duration, timedelta) else duration
+        )
```

Comment:
I suggest moving the conversion part to the `StreamingDataFrame.tumbling_window()` func and accept strictly floats here.
Plus, we need to add some additional validation (e.g. that duration and grace are > 0).